File: BackgroundLocationUpdates/app/src/main/java/com/google/android/gms/location/sample/backgroundlocationupdates/LocationUpdatesService.java
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gms.location.sample.backgroundlocationupdates;

import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.os.Binder;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.text.TextUtils;
import android.util.Log;
import android.util.LogPrinter;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.model.LatLng;
import com.google.gson.Gson;

import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;

/**
 * Interacts with FusedLocationProvider. Responsible for requesting updates, removing updates, and
 * processing location results. This is a bound and started service: when a client binds to the
 * service, it gets location updates in non-batched mode, and when the last client unbinds from the
 * service, it switches to getting updates in batched mode. For this example, only a single client,
 * {@link MainActivity} binds to this service. As a result, as long as {@code MainActivity} is in
 * the foreground, location updates are delivered in real time. When the activity is no longer in
 * the foreground, location updates are batched. When the activity returns to the foreground,
 * batched updates are flushed using
 * {@link com.google.android.gms.location.FusedLocationProviderApi#flushLocations(GoogleApiClient)}.
 *
 */

public class LocationUpdatesService extends Service implements
        GoogleApiClient.OnConnectionFailedListener,
        GoogleApiClient.ConnectionCallbacks {

    private static final String TAG = LocationUpdatesService.class.getSimpleName();

    /**
     * The amount of time to wait to confirm that this service is no longer bound to from a client.
     */
    public static final long ACTIVITY_GONE_CONFIRMATION_DURATION = 10 * 1000; // 10 seconds.

    protected static final String ACTION_INCOMING_LOCATION_UPDATE = BuildConfig.APPLICATION_ID +
            "action.INCOMING_LOCATION_UPDATE";

    // LocationRequest objects.
    private LocationRequest mBatchedLocationRequest;
    private LocationRequest mNonBatchedLocationRequest;

    public LocationUpdatesService() {
    }

    private IBinder mBinder = new MyBinder();

    /**
     * Handler for sending and processing Runnable objects.
     */
    private Handler mServiceHandler;

    /**
     * The entry point to Google Play services.
     */
    private GoogleApiClient mGoogleApiClient;

    /**
     * Helper for tracking whether a calls to unbind are temporary (when the bound client changes
     * orientation, for example), or longer lasting.
     */
    private boolean mActivityReallyGone = false;

    @Override
    public IBinder onBind(Intent intent) {
        Log.i(TAG, "service bound");
        requestNonBatchedUpdates();
        return mBinder;
    }

    @Override
    public void onRebind(Intent intent) {
        super.onRebind(intent);
        Log.i(TAG, "service rebound");
        requestNonBatchedUpdates();
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Log.i(TAG, "Unbinding from service");
        mActivityReallyGone = true;
        // Wait to ensure that this method isn't immediately followed onRebind(). If the binding
        // client is really gone, switch to batched updates.
        mServiceHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                // Check that the activity is still gone. If it is, switch to batched mode.
                if (mActivityReallyGone) {
                    if (Utils.requestingLocationUpdates(getApplicationContext())) {
                        googleApiClientBlockingConnect();
                        LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient,
                                getPendingIntent());
                        Log.i(TAG, "getting batched location updates");
                        LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleApiClient,
                                mBatchedLocationRequest, getPendingIntent());
                    }
                }
            }
        }, ACTIVITY_GONE_CONFIRMATION_DURATION);

        // Returning true allows for onRebind() to run when a client binds again.
        return true;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .addApi(LocationServices.API)
                .build();

        mGoogleApiClient.connect();

        mNonBatchedLocationRequest = Utils.createNonBatchedLocationRequest();
        mBatchedLocationRequest = Utils.createBatchedLocationRequest();

        HandlerThread thread = new HandlerThread(TAG);
        thread.start();
        mServiceHandler = new Handler(thread.getLooper());
    }

    @Override
    public int onStartCommand(final Intent intent, int flags, int startId) {
        final String action = intent.getAction();

        // Process incoming location updates.
        if (TextUtils.equals(ACTION_INCOMING_LOCATION_UPDATE, action)) {
            mServiceHandler.post(new Runnable() {
                @Override
                public void run() {
                    LocationResult result = LocationResult.extractResult(intent);
                    if (result != null) {
                        List<Location> locations = result.getLocations();
                        Gson gson = new Gson();
                        ArrayList<LatLng> allLatLngs = new ArrayList<>();
                        SharedPreferences sp = getSharedPreferences(BuildConfig.APPLICATION_ID,
                                Context.MODE_PRIVATE);
                        String previousLatLngString = sp.getString(Constants.KEY_LOCATION_DATA, "");
                        if (!TextUtils.isEmpty(previousLatLngString)) {
                            ArrayList<LatLng> previousLatLngArrayList = new Gson().fromJson(
                                    previousLatLngString,
                                    new TypeToken<ArrayList<LatLng>>() {
                                    }.getType());
                            allLatLngs.addAll(previousLatLngArrayList);
                        }
                        for (Location location : locations) {
                            location.dump(new LogPrinter(Log.DEBUG, TAG), TAG);
                            allLatLngs.add(new LatLng(location.getLatitude(),
                                    location.getLongitude()));
                        }
                        sp.edit().putString(Constants.KEY_LOCATION_DATA,
                                gson.toJson(allLatLngs)).apply();
                    }
                }
            });
        }
        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "destroying service");
        mServiceHandler.post(new Runnable() {
            @Override
            public void run() {
                googleApiClientBlockingConnect();
                LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient,
                        getPendingIntent());
                mGoogleApiClient.disconnect();
            }
        });
        super.onDestroy();
    }


    @Override
    public void onConnected(Bundle bundle) {
        Log.i(TAG, "GoogleApiClient onConnected");
    }

    @Override
    public void onConnectionSuspended(int i) {
        Log.i(TAG, "GoogleApiClient connection suspended");
    }

    @Override
    public void onConnectionFailed(ConnectionResult connectionResult) {
        Log.i(TAG, "GoogleApiClient connection failed");
    }

    /**
     * Returns the PendingIntent associated with this service. Used when requesting or removing
     * location updates.
     */
    private PendingIntent getPendingIntent() {
        Intent intent = new Intent(this, this.getClass());
        intent.setAction(ACTION_INCOMING_LOCATION_UPDATE);
        return PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    protected void requestNonBatchedUpdates() {
        mActivityReallyGone = false;
        if (Utils.requestingLocationUpdates(getApplicationContext())) {
            mServiceHandler.post(new Runnable() {
                @Override
                public void run() {
                    googleApiClientBlockingConnect();
                    // We may have been getting batched updates before. We flush batched locations
                    // and switch to non-batch mode.
                    LocationServices.FusedLocationApi.flushLocations(mGoogleApiClient);
                    LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient,
                            getPendingIntent());

                    Log.i(TAG, "getting non-batched location updates");
                    LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleApiClient,
                            mNonBatchedLocationRequest, getPendingIntent());
                }
            });
        }
    }

    protected void removeUpdates() {
        mServiceHandler.post(new Runnable() {
            @Override
            public void run() {
                googleApiClientBlockingConnect();
                LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient,
                        getPendingIntent());
                Log.i(TAG, "stopping service");
                stopSelf();
            }
        });
    }

    private void googleApiClientBlockingConnect() {
        if (!mGoogleApiClient.isConnected()) {
            mGoogleApiClient.blockingConnect();
        }
    }

    public class MyBinder extends Binder {
        LocationUpdatesService getService() {
            return LocationUpdatesService.this;
        }
    }
}

File: LocationUpdatesForegroundServiceKotlin/app/src/main/java/com/google/android/gms/location/sample/locationupdatesforegroundservice/LocationUpdatesService.kt
/**
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.gms.location.sample.locationupdatesforegroundservice

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.location.Location
import android.os.*
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.gms.location.*
import com.google.android.gms.tasks.Task

/**
 * A bound and started service that is promoted to a foreground service when location updates have
 * been requested and all clients unbind.
 *
 * For apps running in the background on "O" devices, location is computed only once every 10
 * minutes and delivered batched every 30 minutes. This restriction applies even to apps
 * targeting "N" or lower which are run on "O" devices.
 *
 * This sample show how to use a long-running service for location updates. When an activity is
 * bound to this service, frequent location updates are permitted. When the activity is removed
 * from the foreground, the service promotes itself to a foreground service, and location updates
 * continue. When the activity comes back to the foreground, the foreground service stops, and the
 * notification associated with that service is removed.
 */
class LocationUpdatesService : Service() {
    private val mBinder: IBinder = LocalBinder()
    
    /**
     * Used to check whether the bound activity has really gone away and not unbound as part of an
     * orientation change. We create a foreground service notification only if the former takes
     * place.
     */
    private var mChangingConfiguration = false
    private var mNotificationManager: NotificationManager? = null
    
    /**
     * Contains parameters used by [com.google.android.gms.location.FusedLocationProviderClient].
     */
    private var mLocationRequest: LocationRequest? = null
    
    /**
     * Provides access to the Fused Location Provider API.
     */
    private var mFusedLocationClient: FusedLocationProviderClient? = null
    
    /**
     * Callback for changes in location.
     */
    private var mLocationCallback: LocationCallback? = null
    private var mServiceHandler: Handler? = null
    
    /**
     * The current location.
     */
    private var mLocation: Location? = null
    override fun onCreate() {
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        mLocationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult) {
                super.onLocationResult(locationResult)
                onNewLocation(locationResult.lastLocation)
            }
        }
        createLocationRequest()
        lastLocation
        val handlerThread = HandlerThread(TAG)
        handlerThread.start()
        mServiceHandler = Handler(handlerThread.looper)
        mNotificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        
        // Android O requires a Notification Channel.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name: CharSequence = getString(R.string.app_name)
            // Create the channel for the notification
            val mChannel = NotificationChannel(CHANNEL_ID, name, NotificationManager.IMPORTANCE_DEFAULT)
            
            // Set the Notification Channel for the Notification Manager.
            mNotificationManager!!.createNotificationChannel(mChannel)
        }
    }
    
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Log.i(TAG, "Service started")
        val startedFromNotification = intent.getBooleanExtra(EXTRA_STARTED_FROM_NOTIFICATION,
            false)
        
        // We got here because the user decided to remove location updates from the notification.
        if (startedFromNotification) {
            removeLocationUpdates()
            stopSelf()
        }
        // Tells the system to not try to recreate the service after it has been killed.
        return START_NOT_STICKY
    }
    
    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        mChangingConfiguration = true
    }
    
    override fun onBind(intent: Intent): IBinder? {
        // Called when a client (MainActivity in case of this sample) comes to the foreground
        // and binds with this service. The service should cease to be a foreground service
        // when that happens.
        Log.i(TAG, "in onBind()")
        stopForeground(true)
        mChangingConfiguration = false
        return mBinder
    }
    
    override fun onRebind(intent: Intent) {
        // Called when a client (MainActivity in case of this sample) returns to the foreground
        // and binds once again with this service. The service should cease to be a foreground
        // service when that happens.
        Log.i(TAG, "in onRebind()")
        stopForeground(true)
        mChangingConfiguration = false
        super.onRebind(intent)
    }
    
    override fun onUnbind(intent: Intent): Boolean {
        Log.i(TAG, "Last client unbound from service")
        
        // Called when the last client (MainActivity in case of this sample) unbinds from this
        // service. If this method is called due to a configuration change in MainActivity, we
        // do nothing. Otherwise, we make this service a foreground service.
        if (!mChangingConfiguration && Utils.requestingLocationUpdates(this)) {
            Log.i(TAG, "Starting foreground service")
            startForeground(NOTIFICATION_ID, notification)
        }
        return true // Ensures onRebind() is called when a client re-binds.
    }
    
    override fun onDestroy() {
        mServiceHandler!!.removeCallbacksAndMessages(null)
    }
    
    /**
     * Makes a request for location updates. Note that in this sample we merely log the
     * [SecurityException].
     */
    fun requestLocationUpdates() {
        Log.i(TAG, "Requesting location updates")
        Utils.setRequestingLocationUpdates(this, true)
        startService(Intent(applicationContext, LocationUpdatesService::class.java))
        try {
            mFusedLocationClient!!.requestLocationUpdates(mLocationRequest,
                mLocationCallback, Looper.myLooper())
        } catch (unlikely: SecurityException) {
            Utils.setRequestingLocationUpdates(this, false)
            Log.e(TAG, "Lost location permission. Could not request updates. $unlikely")
        }
    }
    
    /**
     * Removes location updates. Note that in this sample we merely log the
     * [SecurityException].
     */
    fun removeLocationUpdates() {
        Log.i(TAG, "Removing location updates")
        try {
            mFusedLocationClient!!.removeLocationUpdates(mLocationCallback)
            Utils.setRequestingLocationUpdates(this, false)
            stopSelf()
        } catch (unlikely: SecurityException) {
            Utils.setRequestingLocationUpdates(this, true)
            Log.e(TAG, "Lost location permission. Could not remove updates. $unlikely")
        }
    }// Builder initialization for other versions of Android// Extra to help us figure out if we arrived in onStartCommand via the notification or not.
    
    // The PendingIntent that leads to a call to onStartCommand() in this service.
    
    // The PendingIntent to launch activity.
    
    // Builder initialization for Android O.
    /**
     * Returns the [NotificationCompat] used as part of the foreground service.
     */
    private val notification: Notification
        private get() {
            val intent = Intent(this, LocationUpdatesService::class.java)
            val text: CharSequence = Utils.getLocationText(mLocation)
            
            // Extra to help us figure out if we arrived in onStartCommand via the notification or not.
            intent.putExtra(EXTRA_STARTED_FROM_NOTIFICATION, true)
            
            // The PendingIntent that leads to a call to onStartCommand() in this service.
            val servicePendingIntent = PendingIntent.getService(this, 0, intent,
                PendingIntent.FLAG_UPDATE_CURRENT)
            
            // The PendingIntent to launch activity.
            val activityPendingIntent = PendingIntent.getActivity(this, 0,
                Intent(this, MainActivity::class.java), 0)
            val builder: NotificationCompat.Builder
            
            // Builder initialization for Android O.
            builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                NotificationCompat.Builder(
                    this, CHANNEL_ID
                )
                // Builder initialization for other versions of Android
            } else {
                NotificationCompat.Builder(
                    this, ""
                )
            }
            builder.addAction(R.drawable.ic_launch, getString(R.string.launch_activity),
                activityPendingIntent)
            builder.addAction(R.drawable.ic_cancel, getString(R.string.remove_location_updates),
                servicePendingIntent)
            builder.setContentText(text)
            builder.setContentTitle(Utils.getLocationTitle(this))
            builder.setOngoing(true)
            builder.priority = NotificationCompat.PRIORITY_HIGH
            builder.setSmallIcon(R.mipmap.ic_launcher)
            builder.setTicker(text)
            builder.setWhen(System.currentTimeMillis())
            return builder.build()
        }
    private val lastLocation: Unit
        private get() {
            try {
                mFusedLocationClient!!.lastLocation
                    .addOnCompleteListener { task: Task<Location?> ->
                        if (task.isSuccessful && task.result != null) {
                            mLocation = task.result
                        } else {
                            Log.w(TAG, "Failed to get location.")
                        }
                    }
            } catch (unlikely: SecurityException) {
                Log.e(TAG, "Lost location permission.$unlikely")
            }
        }
    
    private fun onNewLocation(location: Location) {
        Log.i(TAG, "New location: $location")
        mLocation = location
        
        // Notify anyone listening for broadcasts about the new location.
        val intent = Intent(ACTION_BROADCAST)
        intent.putExtra(EXTRA_LOCATION, location)
        LocalBroadcastManager.getInstance(applicationContext).sendBroadcast(intent)
        
        // Update notification content if running as a foreground service.
        if (serviceIsRunningInForeground(this)) {
            mNotificationManager!!.notify(NOTIFICATION_ID, notification)
        }
    }
    
    /**
     * Sets the location request parameters.
     */
    private fun createLocationRequest() {
        mLocationRequest = LocationRequest()
        mLocationRequest!!.interval = UPDATE_INTERVAL_IN_MILLISECONDS
        mLocationRequest!!.fastestInterval = FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS
        mLocationRequest!!.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
    }
    
    /**
     * Class used for the client Binder.  Since this service runs in the same process as its
     * clients, we don't need to deal with IPC.
     */
    inner class LocalBinder : Binder() {
        val service: LocationUpdatesService
            get() = this@LocationUpdatesService
    }
    
    /**
     * Returns true if this is a foreground service.
     *
     * @param context The [Context].
     */
    fun serviceIsRunningInForeground(context: Context): Boolean {
        val manager = context.getSystemService(
            ACTIVITY_SERVICE) as ActivityManager
        for (service in manager.getRunningServices(Int.MAX_VALUE)) {
            if (javaClass.name == service.service.className) {
                if (service.foreground) {
                    return true
                }
            }
        }
        return false
    }
    
    companion object {
        private const val PACKAGE_NAME = "com.google.android.gms.location.sample.locationupdatesforegroundservice"
        private val TAG = LocationUpdatesService::class.java.simpleName
        
        /**
         * The name of the channel for notifications.
         */
        private const val CHANNEL_ID = "channel_01"
        const val ACTION_BROADCAST = PACKAGE_NAME + ".broadcast"
        const val EXTRA_LOCATION = PACKAGE_NAME + ".location"
        private const val EXTRA_STARTED_FROM_NOTIFICATION = PACKAGE_NAME +
            ".started_from_notification"
        
        /**
         * The desired interval for location updates. Inexact. Updates may be more or less frequent.
         */
        private const val UPDATE_INTERVAL_IN_MILLISECONDS: Long = 10000
        
        /**
         * The fastest rate for active location updates. Updates will never be more frequent
         * than this value.
         */
        private const val FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = UPDATE_INTERVAL_IN_MILLISECONDS / 2
        
        /**
         * The identifier for the notification displayed for the foreground service.
         */
        private const val NOTIFICATION_ID = 12345678
    }
}